✨ Add interactive stack trace

ryan-doan opened the pull request:
#### Summary:
Stack trace now comes with detailed description of where the program fails. This is also saved in a text file in pros-cli/pros/serial/stack_trace.txt

#### Motivation:
This change help ease the debug of seg faults.

##### References (optional):
Issue #76

#### Test Plan:
The code should be tested with stack traces of varied address count, and compared with output from addr2line with hot package, cold package, and monolith to make sure it doesn't omit important information (the algorithm displays addr2line output that contains the character '?' as "??", since these output are most of the time gibberish that doesn't aid debugging, but I'm not sure if this is the case ONLY for gibberish output, which can lead to the algorithm omitting useful debug information).

- [ ] test item

ryan-doan committed:
Formatted stack trace and save to file

Stack trace now comes with an indicator of which line crashes the program. This is formatted as "cold.package : hot.package : monolith". This trace is also saved to a .txt file

ryan-doan committed:
code formatting changes

ryan-doan committed:
Delete stack_trace.py

This is where I previously had the code for ease of testing. Deleting it now that I've move the code to terminal.py.

Richard-Stump commented:
One thing I noticed when looking through the code is that the paths in the code use `\`, which is a windows only thing. The paths should be changed to use `/` to be platform independent.

ryan-doan committed:
Update pros/serial/terminal/terminal.py

Co-authored-by: BennyBot <48661356+BennyBot@users.noreply.github.com>

ryan-doan committed:
Update pros/cli/terminal.py

Co-authored-by: BennyBot <48661356+BennyBot@users.noreply.github.com>

ryan-doan commented on pros/serial/terminal/terminal.py:
```diff
@@ -225,10 +225,32 @@ def reader(self):
+                                    return ' : ??'
+                                else:
+                                    return ' : ' + temp[15: len(temp)-2]
+                            out += "    " + s + getTrace(s, "..\..\..\\test-project2\\bin\hot.package.elf")
```
if we are expecting the current directory to be the project base folder, is it sufficient to just adjust the path to "bin/package.elf"? Also, I think i can just check for os.path.exists(elf) in getTrace(), and return if the elf file doesn't exists.

ryan-doan committed:
update correct path, prevent addr2line from running if elf file doesn't exists

ryan-doan committed:
Merge branch 'add-interactive-stack-trace' of https://github.com/purduesigbots/pros-cli into add-interactive-stack-trace

ryan-doan committed:
remove project folder

ryan-doan committed:
added css styling to upload button

ryan-doan committed:
Revert "added css styling to upload button"

This reverts commit 059b0773268497a77f8c5e43ed17aa5d491669d0.

ryan-doan committed:
remove stack_trace.txt

ryan-doan committed:
fix errors with path

ryan-doan committed:
rewrote code to make it functional

The previous algorithm does not work with how terminal.py prints the stack trace. This algorithm takes it into account, and has been verified to work with a VEX brain.

BennyBot committed:
Remove sqlalchemy, add options, string formatting, add colored output.

ryan-doan committed:
Merge branch 'develop' into add-interactive-stack-trace

omegaStag commented:
#242

Andrewtho5942 committed:
add shell argument to subprocess open

Fixes a fileNotFound error

Ghurfa committed:
Merging in develop

Ghurfa committed:
Changed the addr2line path to use the toolchain

Rocky14683 committed:
Merge branch 'develop' into add-interactive-stack-trace

Rocky14683 committed:
Finish data abort realtime stack trace functionality

Rocky14683 committed:
Merge remote-tracking branch 'origin/add-interactive-stack-trace' into add-interactive-stack-trace

Rocky14683 committed:
Work on formatting and color

Rocky14683 commented:
Progress:
<img width="768" alt="截圖 2024-07-23 上午1 23 13" src="https://github.com/user-attachments/assets/fed91f09-0bfe-4ceb-8013-5e005bbfb129">
Any thought on indentation and colors?

WillXuCodes commented:
> Progress: <img alt="截圖 2024-07-23 上午1 23 13" width="768" src="https://private-user-images.githubusercontent.com/101498190/351173617-fed91f09-0bfe-4ceb-8013-5e005bbfb129.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjE3MTIzMTUsIm5iZiI6MTcyMTcxMjAxNSwicGF0aCI6Ii8xMDE0OTgxOTAvMzUxMTczNjE3LWZlZDkxZjA5LTBiZmUtNGNlYi04MDEzLTVlMDA1YmJmYjEyOS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNzIzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDcyM1QwNTIwMTVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1kZTc1YjFhZWYzMzNjYTUxNzI1NTU3ZmZhYjM2ZTI5OTMwNTZkNDQ4MjI5ZWY5Nzg5NzcwMjNhOGE5ZWRjYzA3JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.ClGgOOjDNBsNFkLu0WWpQfJ4g5DsI26ugrJE-KnWpjs"> Any thought on indentation and colors?

Wonder how this would look on a small window (like the VSCode integrated terminal). 

One other interesting thought, get rid of the "Registers at Exception?" section? We might want to consider shaving some of the fat that your average user may not need. Most important part is the stack trace section. (I'm fully aware addr2line provides all this data, not feasible how this is since it's been so long :skull: )

ion098 commented on pros/serial/terminal/terminal.py:
```diff
@@ -224,7 +231,75 @@ def reader(self):
+                        def getTrace(s, path):
+                            if not os.path.exists(path):
+                                return ''
+                            temp = subprocess.check_output([addr2line_path, '-faps', '-e', path, s]).decode('utf-8')
```
```suggestion
                            temp = subprocess.check_output([addr2line_path, '-faps', '-e', '-C', path, s]).decode('utf-8')
```
Right now C++ function names look like letter soup, since they're not demangled.
